src/Features/Core/Portable/ValidateFormatString/AbstractValidateFormatStringDiagnosticAnalyzer.cs

```diff
@@ -0,0 +1,403 @@
+            description: Description);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
+            => ImmutableArray.Create(Rule);
+
+        /// <summary>
+        /// this regex is used to remove escaped brackets from
+        /// the format string before looking for valid {} pairs
+        /// </summary>
+        private static Regex s_removeEscapedBracketsRegex = new Regex("{{");
+
+        /// <summary>
+        /// this regex is used to extract the text between the
+        /// brackets and save the contents in a MatchCollection
+        /// </summary>
+        private static Regex s_extractPlaceholdersRegex = new Regex("{(.*?)}");
+
+        private const string NameOfArgsParameter = "args";
+        private const string NameOfFormatStringParameter = "format";
+
+        protected abstract ISyntaxFactsService GetSyntaxFactsService();
+        protected abstract TSyntaxKind GetInvocationExpressionSyntaxKind();
+        protected abstract SyntaxNode GetArgumentExpression(SyntaxNode syntaxNode);
+        protected abstract SyntaxNode TryGetMatchingNamedArgument(SeparatedSyntaxList<SyntaxNode> arguments, string searchArgumentName);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.RegisterCompilationStartAction(startContext =>
+            {
+                var formatProviderType = startContext.Compilation.GetTypeByMetadataName(typeof(System.IFormatProvider).FullName);
+                if (formatProviderType == null)
+                {
+                    return;
+                }
+
+                startContext.RegisterSyntaxNodeAction(
+                    c => AnalyzeNode(c, formatProviderType),
+                    GetInvocationExpressionSyntaxKind());
+            });
+        }
+
+        private void AnalyzeNode(SyntaxNodeAnalysisContext context, INamedTypeSymbol formatProviderType)
+        {
+            var optionSet = context.Options.GetDocumentOptionSetAsync(
+                context.Node.SyntaxTree, context.CancellationToken).GetAwaiter().GetResult();
+
+            if (optionSet.GetOption(
+                ValidateFormatStringOption.ReportInvalidPlaceholdersInStringDotFormatCalls,
+                context.SemanticModel.Language) == false)
+            {
+                return;
+            }
+
+            var syntaxFacts = GetSyntaxFactsService();
+            var expression = syntaxFacts.GetExpressionOfInvocationExpression(context.Node);
+
+            if (!IsValidFormatMethod(syntaxFacts, expression))
+            {
+                return;
+            }
+
+            var arguments = syntaxFacts.GetArgumentsOfInvocationExpression(context.Node);
+            var numberOfArguments = arguments.Count;
+            var symbolInfo = context.SemanticModel.GetSymbolInfo(expression, context.CancellationToken);
+
+            var method = TryGetValidFormatMethodSymbol(context, numberOfArguments, symbolInfo);
```

Comment:
We should make sure that this is a valid format method before enumerating all the arguments.